### PR TITLE
Integration test for the shadow library (device-shadow-for-aws-iot-embedded-sdk)

### PR DESCRIPTION
**Pull request by andysun2015**
Description
-----------

- Port integration test for the shadow library from [aws-iot-device-sdk-embedded-C](https://github.com/aws/aws-iot-device-sdk-embedded-C/).
- This PR duplicates https://github.com/aws/amazon-freertos/pull/2538 and the comments are addressed.

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

### CODE REVIEW

**Review by sarenameas — APPROVED** [review]

File: tests/integration_test/shadow_system_test.c [hunk 1]
```diff
@@ -0,0 +1,912 @@
...
+    socketsConfig.maxFragmentLength = 0;
+    socketsConfig.disableSni = true;
+    socketsConfig.pRootCa = SERVER_ROOT_CA_CERT;
+    socketsConfig.rootCaSize = strlen( SERVER_ROOT_CA_CERT ) + 1U;
+    socketsConfig.sendTimeoutMs = TRANSPORT_SEND_RECV_TIMEOUT_MS;
+    socketsConfig.recvTimeoutMs = TRANSPORT_SEND_RECV_TIMEOUT_MS;
+
+    /* Establish a TCP connection with the server endpoint, then
+     * establish TLS session on top of TCP connection. */
+    TEST_ASSERT_EQUAL( TRANSPORT_SOCKET_STATUS_SUCCESS, SecureSocketsTransport_Connect( &networkContext,
```
Comment (yourslab) [inline tests/integration_test/shadow_system_test.c:770]: Can you please add retry logic for connection here?

**Review by yourslab — APPROVED** [review]

File: tests/integration_test/shadow_system_test.c [hunk 2]
```diff
@@ -0,0 +1,932 @@
...
+/* Include logging header files and define logging macros in the following order:
+ * 1. Include the header file "logging_levels.h".
+ * 2. Define the LIBRARY_LOG_NAME and LIBRARY_LOG_LEVEL macros depending on
+ * the logging configuration for DEMO.
+ * 3. Include the header file "logging_stack.h", if logging is enabled for DEMO.
+ */
+
+#include "logging_levels.h"
+
+/* Logging configuration for the Demo. */
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:75]:
```suggestion
/* Logging configuration for the Tests. */
```

File: tests/integration_test/shadow_system_test.c [hunk 3]
```diff
@@ -0,0 +1,932 @@
...
+#ifndef LIBRARY_LOG_NAME
+    #define LIBRARY_LOG_NAME    "TEST"
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:77]:
```suggestion
    #define LIBRARY_LOG_NAME    "SHADOW_TEST"
```

File: tests/integration_test/shadow_system_test.c [hunk 4]
```diff
@@ -0,0 +1,932 @@
...
+#endif
+
+#ifndef LIBRARY_LOG_LEVEL
+    #define LIBRARY_LOG_LEVEL    LOG_INFO
+#endif
+#include "logging_stack.h"
+
+/**
+ * These configuration settings are required to run the shadow library integration test.
+ * Throw compilation error if the below configs are not defined.
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:87]: Does not seem correct - the code does not throw compiler error but provides default values.

File: tests/integration_test/shadow_system_test.c [hunk 5]
```diff
@@ -0,0 +1,932 @@
...
+    /* Convert the ticks to milliseconds. */
+    timeMs = ( uint32_t ) tickCount * MILLISECONDS_PER_TICK;
+
+    /* Reduce globalEntryTime from obtained time so as to always return the
+     * elapsed time in the application. */
+    timeMs = ( uint32_t ) ( timeMs - globalEntryTime );
+
+    return timeMs;
+}
+
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:359]: Does running uncrustify not add a function separator after each function? 

```
/*-----------------------------------------------------------*/
```

File: tests/integration_test/shadow_system_test.c [hunk 6]
```diff
@@ -0,0 +1,932 @@
...
+                                                 &messageType,
+                                                 &pThingName,
+                                                 &thingNameLength ) )
+        {
+            /* Upon successful return, the messageType has been filled in. */
+            if( messageType == ShadowMessageTypeUpdateDelta )
+            {
+                receivedUpdateDeltaResult = true;
+                LogInfo( ( "/update/delta payload:%s.\n\n", ( const char * ) pDeserializedInfo->pPublishInfo->pPayload ) );
+                /* validate the each field in json payload if meet our expectation. */
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:461]: Where is the validation? Or is this a misplaced comment?

File: tests/integration_test/shadow_system_test.c [hunk 7]
```diff
@@ -0,0 +1,932 @@
...
+
+    /* Start with everything at 0. */
+    ( void ) memset( ( void * ) pSubscriptionList, 0x00, sizeof( pSubscriptionList ) );
+
+    /* topic /update/accepted and uses qos. */
+    pSubscriptionList[ 0 ].qos = qos;
+    pSubscriptionList[ 0 ].pTopicFilter = pTopic;
+    pSubscriptionList[ 0 ].topicFilterLength = topicLength;
+
+    /* Generate packet identifier for the SUBSCRIBE packet. */
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:656]:
```suggestion
    /* Generate packet identifier for the UNSUBSCRIBE packet. */
```

File: tests/integration_test/shadow_system_test.c [hunk 8]
```diff
@@ -0,0 +1,932 @@
...
+/**
+ * @brief Subscribes the shadow topics: /update/delta, /update/documents,
+ * /update/accepted, /delete/accepted, /get/accepted, then publish the
+ * regarding payloads to verify if receiving the notification from the
+ * subscribed topics.
+ */
+TEST( deviceShadow_Integration, test_Shadow_System )
+{
+    /* A buffer containing the update document. It has static duration to prevent
+     * it from being placed on the call stack. */
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:839]:
```suggestion
     * it from being placed on the stack. */
```

File: tests/integration_test/shadow_system_test.c [hunk 9]
```diff
@@ -0,0 +1,932 @@
...
+    static char updateDocument[ 1 ] = { 0 };
```
Comment (aggarg) [inline tests/integration_test/shadow_system_test.c:840]: A better name would be `nullDocument` or `emptyDocument`

File: tests/integration_test/shadow_system_test.c (hunk 4 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:87]: ok~I will make it as follows:
```
/* Default values for configs. */
```

File: tests/integration_test/shadow_system_test.c (hunk 5 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:359]: No, it doesn't add the function separator.

File: tests/integration_test/shadow_system_test.c (hunk 6 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:461]: oh~sorry, it's a misplaced comment.

File: tests/integration_test/shadow_system_test.c (hunk 7 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:656]: Thanks.

File: tests/integration_test/shadow_system_test.c (hunk 8 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:839]: Thanks.

File: tests/integration_test/shadow_system_test.c (hunk 9 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:840]: thanks for the suggestion. I'll take emptyDocument.

File: tests/integration_test/shadow_system_test.c [hunk 10]
```diff
@@ -0,0 +1,942 @@
...
+ */
+#define TEST_CLIENT_IDENTIFIER_LENGTH       ( sizeof( TEST_CLIENT_IDENTIFIER ) - 1U )
+
+/**
+ * @brief Transport timeout in milliseconds for transport send and receive.
+ */
+#define TRANSPORT_SEND_RECV_TIMEOUT_MS      ( 2000U )
+
+/**
+ * @brief Timeout for receiving CONNACK packet in milli seconds.
```
Comment (sarenameas) [inline tests/integration_test/shadow_system_test.c:122]:
```suggestion
 * @brief Timeout for receiving CONNACK packet in milliseconds.
```

File: tests/integration_test/shadow_system_test.c [hunk 11]
```diff
@@ -0,0 +1,942 @@
...
+
+    packetIdentifier = pDeserializedInfo->packetIdentifier;
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        assert( pDeserializedInfo->pPublishInfo != NULL );
+        /* Handle incoming publish. */
```
Comment (sarenameas) [inline tests/integration_test/shadow_system_test.c:447]: Can you delete this duplicate comment?

File: tests/integration_test/shadow_system_test.c [hunk 12]
```diff
@@ -0,0 +1,942 @@
...
+    /* Initializer server information. */
+    serverInfo.pHostName = BROKER_ENDPOINT;
+    serverInfo.hostNameLength = strlen( BROKER_ENDPOINT );
+    serverInfo.port = BROKER_PORT;
+
+    /* Initialize SocketsConfig. */
+    socketsConfig.enableTls = true;
+    socketsConfig.pAlpnProtos = NULL;
+    socketsConfig.maxFragmentLength = 0;
+    socketsConfig.disableSni = true;
```
Comment (sarenameas) [inline tests/integration_test/shadow_system_test.c:744]: Please set to this false. We are connecting to AWS IoT.

Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:744]: ok~

File: tests/integration_test/shadow_system_test.c (hunk 11 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:447]: thanks for the reminder.

File: tests/integration_test/shadow_system_test.c (hunk 10 above)
Comment (andysun2015) [reply tests/integration_test/shadow_system_test.c:122]: thanks.

**Review by aggarg — APPROVED** [review]

**Review by sarenameas — APPROVED** [review]

**Pull request merged**